plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -177,7 +176,10 @@ private Stream<DeltaLakeSplit> getSplits(
 
         MetadataEntry metadataEntry = tableHandle.getMetadataEntry();
         boolean isOptimize = tableHandle.isOptimize();
-        Set<Map<String, Optional<String>>> partitionsWithAtMostOneFile = isOptimize ? findPartitionsWithAtMostOneFile(validDataFiles) : ImmutableSet.of();
+        if (isOptimize && maxScannedFileSizeInBytes.isPresent()) {
+            validDataFiles = filterValidDataFilesForOptimize(validDataFiles, maxScannedFileSizeInBytes.get())
```

Comment:
why would we want to do this only when maxScannedFileSizeInBytes is present?

if there is just one file per partition, no need to run optimize over those partitions, regardless of the file size.

Reply:
> why would we want to do this only when maxScannedFileSizeInBytes is present?

I took over the initial logic from https://github.com/trinodb/trino/pull/18938 - I'm guessing though that `maxScannedFileSizeInBytes` must actually be present when doing `OPTIMIZE`. Will add a check for presence.

> if there is just one file per partition, no need to run optimize over those partitions, regardless of the file size.

It's a bit more nuanced than "just one file per partition" - see https://github.com/trinodb/trino/issues/20088
We have to disregard for optimization files which are already big.

Reply:
I'm extracting the 1 commit & the fix for #20088 to a separate PR